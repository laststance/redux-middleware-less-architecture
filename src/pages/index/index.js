// @flow
import React from 'react'
import styled from 'styled-components'
import { Link } from 'react-router-dom'
import { flexRowCenter } from '../../const'

const Container = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
`
const TopContainer = styled.div`
  height: 50%;
  ${flexRowCenter};
`
const BottomContainer = styled.div`
  height: 50%;
  ${flexRowCenter};
`

// TODO Error Bound
const index = () => {
  return (
    <Container>
      <TopContainer>
        <h1>Index Page</h1>
      </TopContainer>
      <BottomContainer>
        <Link to="/github">GitHub Page</Link>
      </BottomContainer>
    </Container>
  )
}

export default index
