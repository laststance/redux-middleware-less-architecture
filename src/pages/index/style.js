import styled from 'styled-components'
import { flexRowCenter } from '../../const'

export const Container = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
`
export const Head = styled.div`
  height: 20%;
  ${flexRowCenter};
`
export const Github = styled.div`
  height: 20%;
  ${flexRowCenter};
`
