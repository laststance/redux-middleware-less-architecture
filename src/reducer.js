// @flow
import type { ReduxAction } from './action'
import type { Reducer } from 'redux'
import type { RepositoryList } from './types/APIDataModel'

// react-router
export type Location = {
  pathname: string,
  search: string,
  hash: string
}

export type AppState = {
  boot: string,
  isLoading: boolean,
  repositoryList: RepositoryList
}

export type RootReduxState = {
  app: AppState,
  router: {
    location: Location
  }
}

export const initialState: AppState = {
  boot: 'uninitialized.',
  isLoading: false,
  repositoryList: []
}

const appReducer: Reducer<AppState, ReduxAction> = (
  state: AppState = initialState,
  action: ReduxAction
): AppState => {
  switch (action.type) {
    case '@@/App/INIT':
      return { ...state, boot: 'initialized by redux.' }

    case 'START_LOADING':
      return { ...state, isLoading: true }

    case 'FETCH_REPOSITORY':
      return {
        ...state,
        repositoryList: action.payload.repositoryList,
        isLoading: false
      }

    default:
      return state
  }
}

export default appReducer
