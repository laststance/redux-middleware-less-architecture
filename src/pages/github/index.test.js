import React, { Component } from 'react'
import { render, wait } from 'react-testing-library'
import 'jest-dom/extend-expect'
import axios from 'axios'
import { Github } from './index'
import mockResponse from '../../testutil/mockResponse'
import { createStore } from 'redux'
import reducer from '../../reducer'

// create redux
const store = createStore(reducer)
class DependencyInjectionCompoment extends Component {
  state = { reduxState: store.getState() }

  constructor(props) {
    super(props)
    store.subscribe(() => this.setState({ reduxState: store.getState() }))
  }

  render() {
    return (
      <Github
        isLoading={this.state.reduxState.isLoading}
        repositoryList={this.state.reduxState.repositoryList}
        dispatch={store.dispatch}
      />
    )
  }
}

describe('github page', () => {
  it('fetch from API data shown when mounted', async () => {
    // Mock axios.get() for disable real http request and set fake responce
    axios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: mockResponse
      })
    )

    // Before mount, isLoading should be false
    expect(store.getState().isLoading).toBe(false)

    const { getByTestId, container } = render(<DependencyInjectionCompoment />)

    // After mount, isLoading should be true
    expect(store.getState().isLoading).toBe(true)
    expect(getByTestId('loading')).toBeInTheDocument()
    expect(getByTestId('github-header')).toHaveTextContent(
      'GitHub Repo Search Example'
    )
    expect(axios.get).toHaveBeenCalledTimes(1)
    expect(axios.get).toHaveBeenCalledWith(
      'https://api.github.com/search/repositories?q=react'
    )

    // Simurate duration of api request
    await wait()

    // Testing what is showing browser, after completed all processes
    expect(store.getState().isLoading).toBe(false)
    const list = getByTestId('repo-list')
    expect(list.children.length).toBe(30)
    expect(container.firstChild).toMatchSnapshot()
  })
})
