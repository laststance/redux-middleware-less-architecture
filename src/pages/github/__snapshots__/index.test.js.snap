// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github page Testing data fetch and display 1`] = `
<div
  class="sc-bdVaJa fKEnoH"
>
  <h1
    class="sc-bwzfXH bDlgwH"
    data-testid="github-header"
  >
    GitHub Repo Search Example
  </h1>
  <div
    class="sc-htpNat KtFsv"
  />
  <div
    class="sc-ifAKCX SdvUM"
    data-testid="repo-list"
  >
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        A declarative, efficient, and flexible JavaScript library for building user interfaces.
      </p>
      <p>
        facebook/react
      </p>
      <p>
        facebook
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/69631?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        Event-driven, non-blocking I/O with PHP.
      </p>
      <p>
        reactphp/react
      </p>
      <p>
        reactphp
      </p>
      <img
        alt="avatar"
        src="https://avatars2.githubusercontent.com/u/1696866?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        一个react+redux+webpack+ES6+antd的SPA的后台管理底层框架
      </p>
      <p>
        duxianwei520/react
      </p>
      <p>
        duxianwei520
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/3249653?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        React docs in Chinese | React 中文文档翻译
      </p>
      <p>
        discountry/react
      </p>
      <p>
        discountry
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/4507101?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        ReactiveCocoa
      </p>
      <p>
        Streams of values over time
      </p>
      <p>
        ReactiveCocoa/ReactiveCocoa
      </p>
      <p>
        ReactiveCocoa
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/3422977?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        京东首页构建
      </p>
      <p>
        Cathy0807/react
      </p>
      <p>
        Cathy0807
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/20653643?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-native
      </p>
      <p>
        A framework for building native apps with React.
      </p>
      <p>
        facebook/react-native
      </p>
      <p>
        facebook
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/69631?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-router
      </p>
      <p>
        Declarative routing for React
      </p>
      <p>
        ReactTraining/react-router
      </p>
      <p>
        ReactTraining
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/11823761?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        基于react的企业后台管理开发框架
      </p>
      <p>
        react-redux-antd-es6/react
      </p>
      <p>
        react-redux-antd-es6
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/24805142?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        create-react-app
      </p>
      <p>
        Create React apps with no build configuration.
      </p>
      <p>
        facebook/create-react-app
      </p>
      <p>
        facebook
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/69631?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        Examples for the React Quickly book. 
      </p>
      <p>
        azat-co/react
      </p>
      <p>
        azat-co
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/1137314?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        storybook
      </p>
      <p>
        Interactive UI component dev & test: React, React Native, Vue, Angular
      </p>
      <p>
        storybooks/storybook
      </p>
      <p>
        storybooks
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/22632046?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-redux-starter-kit
      </p>
      <p>
        Get started with React, Redux, and React-Router.
      </p>
      <p>
        davezuko/react-redux-starter-kit
      </p>
      <p>
        davezuko
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/6439050?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-demos
      </p>
      <p>
        a collection of simple demos of React.js
      </p>
      <p>
        ruanyf/react-demos
      </p>
      <p>
        ruanyf
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/905434?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        Extremely simple boilerplate, easiest you can find, for React application including all the necessary tools: Flow | React 16 | redux | babel 6 | webpack 3 | css-modules | jest | enzyme | express + optional: sass/scss
      </p>
      <p>
        werein/react
      </p>
      <p>
        werein
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/3116920?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        awesome-react
      </p>
      <p>
        A collection of awesome things regarding React ecosystem.
      </p>
      <p>
        enaqx/awesome-react
      </p>
      <p>
        enaqx
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/182219?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-boilerplate
      </p>
      <p>
        :fire: A highly scalable, offline-first foundation with the best developer experience and a focus on performance and best practices.
      </p>
      <p>
        react-boilerplate/react-boilerplate
      </p>
      <p>
        react-boilerplate
      </p>
      <img
        alt="avatar"
        src="https://avatars2.githubusercontent.com/u/25323389?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        ReactiveUI
      </p>
      <p>
        An advanced, composable, functional reactive model-view-viewmodel framework for all .NET platforms that is inspired by functional reactive programming. ReactiveUI allows you to  abstract mutable state away from your user interfaces, express the idea around a feature in one readable place and improve the testability of your application.
      </p>
      <p>
        reactiveui/ReactiveUI
      </p>
      <p>
        reactiveui
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/2327577?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react
      </p>
      <p>
        🔼 UI-Router for React
      </p>
      <p>
        ui-router/react
      </p>
      <p>
        ui-router
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/16856735?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        reactjs101
      </p>
      <p>
        從零開始學 ReactJS（ReactJS 101）是一本希望讓初學者一看就懂的 ReactJS 中文入門教學書，由淺入深學習 ReactJS 生態系 (Flux, Redux, React Router, ImmutableJS, React Native, Relay/GraphQL etc.)。
      </p>
      <p>
        kdchang/reactjs101
      </p>
      <p>
        kdchang
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/2097473?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-select
      </p>
      <p>
        A Select control built with and for React JS
      </p>
      <p>
        JedWatson/react-select
      </p>
      <p>
        JedWatson
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/872310?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-navigation
      </p>
      <p>
        Routing and navigation for your React Native apps
      </p>
      <p>
        react-navigation/react-navigation
      </p>
      <p>
        react-navigation
      </p>
      <img
        alt="avatar"
        src="https://avatars2.githubusercontent.com/u/29647600?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        reaction
      </p>
      <p>
        Reaction is a customizable, real-time reactive, JavaScript commerce platform.
      </p>
      <p>
        reactioncommerce/reaction
      </p>
      <p>
        reactioncommerce
      </p>
      <img
        alt="avatar"
        src="https://avatars0.githubusercontent.com/u/5605462?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-bootstrap
      </p>
      <p>
        Bootstrap 3 components built with React
      </p>
      <p>
        react-bootstrap/react-bootstrap
      </p>
      <p>
        react-bootstrap
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/6853419?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        reactGo
      </p>
      <p>
        Your One-Stop solution for a full-stack universal Redux App!
      </p>
      <p>
        reactGo/reactGo
      </p>
      <p>
        reactGo
      </p>
      <img
        alt="avatar"
        src="https://avatars2.githubusercontent.com/u/20782186?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-redux
      </p>
      <p>
        Official React bindings for Redux
      </p>
      <p>
        reactjs/react-redux
      </p>
      <p>
        reactjs
      </p>
      <img
        alt="avatar"
        src="https://avatars3.githubusercontent.com/u/6412038?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        react-starter-kit
      </p>
      <p>
        React Starter Kit — isomorphic web app boilerplate (Node.js, Express, GraphQL, React.js, Babel, PostCSS, Webpack, Browsersync)
      </p>
      <p>
        kriasoft/react-starter-kit
      </p>
      <p>
        kriasoft
      </p>
      <img
        alt="avatar"
        src="https://avatars2.githubusercontent.com/u/773036?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        ReactNativeReduxCasts
      </p>
      <p>
        Companion Repo for The Complete React Native and Redux Course
      </p>
      <p>
        StephenGrider/ReactNativeReduxCasts
      </p>
      <p>
        StephenGrider
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/5003903?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        React-For-Beginners-Starter-Files
      </p>
      <p>
        Starter files for learning React.js with React for Beginners
      </p>
      <p>
        wesbos/React-For-Beginners-Starter-Files
      </p>
      <p>
        wesbos
      </p>
      <img
        alt="avatar"
        src="https://avatars2.githubusercontent.com/u/176013?v=4"
      />
    </div>
    <div
      class="sc-bxivhb bUZGRr"
    >
      <p>
        React
      </p>
      <p>
        写写React组件
      </p>
      <p>
        leilihuang/React
      </p>
      <p>
        leilihuang
      </p>
      <img
        alt="avatar"
        src="https://avatars1.githubusercontent.com/u/13639327?v=4"
      />
    </div>
  </div>
</div>
`;
